visualizer/draw/area-chart.js

```diff
@@ -150,17 +152,22 @@ class AreaChart extends HtmlContent {
     }
   }
   initializeFromData () {
-    // Same behaviour on mouse movements off coloured area as on
-    this.d3AreaChartSVG.on('mousemove', () => {
-      this.showSlice(d3.event)
-    })
+    if (!this.initialized) {
+      // These actions aren't redone when data changes, but must be done after data was set
+      this.initialized = true
 
-    this.d3ContentWrapper.on('mouseleave', () => {
-      this.hoverBox.hide()
-      this.d3SliceHighlight.classed('hidden', true)
-    })
+      // Same behaviour on mouse movements off coloured area as on
+      this.d3AreaChartSVG.on('mousemove', () => {
+        this.showSlice(d3.event)
+      })
 
+      this.d3ContentWrapper.on('mouseleave', () => {
+        this.hoverBox.hide()
+        this.d3SliceHighlight.classed('hidden', true)
+      })
+    }
     if (this.contentProperties.static) this.d3LeadInText.html(this.getLeadInText())
+    if (!this.d3AreaPaths) this.createPathsForLayout()
```

Comment:
Yes, so the intended pattern is:
- `initializeElements()` is called before `setData` and creates DOM elements, stores D3 selections and binds event listeners as early as possible after page load, creating frames etc that the user sees while the load indicator spins and the data is crunched
- `initializeFromData()` is called after `setData` and contains DOM-manipulating actions that need to happen after the data is available (usually, because the number of DOM nodes to create depends on the data; here, it's because we're binding event handlers that call methods that require data access, so those events shouldn't be handled before the data is available)
- `setData()` itself is called when the `layout` or `dataSet` changes and sets properties of the class from the latest data (and usually calls a `initializeFromData()` if there is one)

So we call it once on page load here, the first time setData() is called. We don't call it every time `setData()` gives us a new layout object, because we know that we don't need to recreate the paths every time a window resize or similar recalculates the layout positions. We only need to change those if the topmost UI changes, so we call that in the topmost UI listener (which isn't called on initial page load).

Reply:
That definitely connected some dots for me, thank you!